### PR TITLE
Update auto-opt docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Updated the docs for automatic_optimization. It's a LightningModule property now. Added docs for a few more properties. Also, update the automatic_optimization code wherever it's required. Other whitespace changes are made by pre-commit.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/common/lightning_module.rst [hunk 1]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
 
 --------------
 
+automatic_optimization
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:935]: I am confused. Shouldn't it be manual optimisation ?

File: docs/source/common/lightning_module.rst [hunk 2]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
...
+
+        gen_loss = ...
+        self.manual_backward(gen_loss, opt_a)
+        opt_a.step()
+        opt_a.zero_grad()
+
+        disc_loss = ...
+        self.manual_backward(disc_loss, opt_b)
+        opt_b.step()
+        opt_b.zero_grad()
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:976]:
```suggestion
        gen_loss = ...
        opt_a.zero_grad()
        self.manual_backward(gen_loss, opt_a)
        opt_a.step()

        disc_loss = ...
        opt_b.zero_grad()
        self.manual_backward(disc_loss, opt_b)
        opt_b.step()
```

File: docs/source/common/lightning_module.rst [hunk 3]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
...
+example_input_array
+~~~~~~~~~~~~~~~~~~~
+Set and access example_input_array which is basically a single batch.
+
+.. code-block:: python
+
+    def __init__(self):
+        self.example_input_array = ...
+
+    def training_step(...):
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:989]: That's a bit weird to have it in training_step. Might confused users.

File: docs/source/common/lightning_module.rst [hunk 4]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
...
+        res = self(self.example_input_array)
+
+--------------
+
+datamodule
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:994]: I find this part really confusing. Looks like you are setting datamodule in LightningModule. I think it is bad practice.

Worse case, for configure_optimizers, `num_trianing_samples = self.trainer.datamodule.train_dataloader()`

File: docs/source/common/lightning_module.rst [hunk 5]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
...
+
+--------------
+
+model_size
+~~~~~~~~~~
+Get the model file size.
+
+.. code-block:: python
+
+    def training_step(...):
```
Comment (tchaton) [inline docs/source/common/lightning_module.rst:1014]: Not sure why model_size should be used in `training_step`.

File: docs/source/common/lightning_module.rst [hunk 6]
```diff
@@ -932,9 +932,93 @@ True if using TPUs
...
+AND you are an expert user. Most useful for research like RL, sparse coding and GAN research.
+
+In the multi-optimizer case, ignore the optimizer_idx flag and use the optimizers directly
+
+.. code-block:: python
+
+    def __init__(self):
+        self.automatic_optimization = False
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
```
Comment (awaelchli) [inline docs/source/common/lightning_module.rst:964]: we should probably also make it so that the opt idx is not even part of the signature when doing manual opt, right? or perhaps it is already like this and just outdated docs.

File: README.md [hunk 7]
```diff
@@ -315,6 +315,10 @@ For complex/professional level work, you have optional full control of the train
 
 ```python
 class LitAutoEncoder(pl.LightningModule):
+    def __init__(self):
+        super().__init__()
+        self.automatic_optimization = False
+
     def training_step(self, batch, batch_idx, optimizer_idx):
         # access your optimizers with use_pl_optimizer=False. Default is True
```
Comment (awaelchli) [inline README.md:323]: this comment has to be removed, right @tchaton?

File: docs/source/common/lightning_module.rst (hunk 1 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:935]: the property is called `automatic_optimization`

File: docs/source/common/lightning_module.rst (hunk 2 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:976]: you sure about this? I just checked a few tests related to this in there it's like:
```python
manual_backward()
step()
zero_grad()
```

File: docs/source/common/lightning_module.rst (hunk 5 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:1014]: ok i'll just remove the example since I'm not sure about it's use-case.

File: docs/source/common/lightning_module.rst (hunk 6 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:964]: it's not like that in the code. Agreed opt_idx doesn't make sense in manual_optimization. will update.

File: README.md (hunk 7 above)
Comment (rohitgr7) [reply README.md:323]: I think the only difference here is if it's true then lightning will handle zero_grad + some hooks + optimizer_step in different backends else it's just a normal pytorch optimizer. @tchaton can you confirm?

File: docs/source/common/lightning_module.rst (hunk 4 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:994]: updated

File: docs/source/common/lightning_module.rst (hunk 3 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:989]: updated

File: docs/source/common/lightning_module.rst [hunk 8]
```diff
@@ -841,7 +841,7 @@ The current step (does not reset each epoch)
 
 hparams
 ~~~~~~~
-After calling `save_hyperparameters` anything passed to init() is available via hparams.
+After calling ``save_hyperparameters`` anything passed to init() is available via hparams.
```
Comment (SkafteNicki) [inline docs/source/common/lightning_module.rst:844]:
```suggestion
After calling ``save_hyperparameters`` anything passed to ``__init__()`` is available via the ``hparams`` attribute.
```

File: docs/source/common/lightning_module.rst [hunk 9]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
 
 --------------
 
+automatic_optimization
+~~~~~~~~~~~~~~~~~~~~~~
+When set to False, Lightning does not automate the optimization process. This means you are responsible for your own
```
Comment (SkafteNicki) [inline docs/source/common/lightning_module.rst:937]:
```suggestion
When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for your own
```

File: docs/source/common/lightning_module.rst [hunk 10]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
...
+        # access your optimizers with use_pl_optimizer=False. Default is True
+        opt = self.optimizers(use_pl_optimizer=True)
+
+        loss = ...
+        self.manual_backward(loss, opt)
+        opt.step()
+        opt.zero_grad()
+
+This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers
+AND you are an expert user. Most useful for research like RL, sparse coding and GAN research.
```
Comment (SkafteNicki) [inline docs/source/common/lightning_module.rst:955]: Spell it out?
```suggestion
AND you are an expert user. Most useful for research like reinforcement learning, sparse coding and GAN research.
```

File: docs/source/starter/new-project.rst [hunk 11]
```diff
@@ -258,16 +258,13 @@ Manual optimization
...
 or an inner loop, you can turn off automatic optimization and fully control the training loop yourself.
 
-First, turn off automatic optimization:
-
-.. testcode::
-
-    trainer = Trainer(automatic_optimization=False)
-
-Now you own the train loop!
+Turn off automatic optimization and you own the train loop!
```
Comment (SkafteNicki) [inline docs/source/starter/new-project.rst:261]: just personal taste
```suggestion
Turn off automatic optimization and you control the train loop!
```

File: docs/source/common/lightning_module.rst [hunk 12]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
...
+
+    def configure_optimizers(self):
+        num_training_samples = len(self.datamodule.train_dataloader())
+        ...
+
+--------------
+
+model_size
+~~~~~~~~~~
+Get the model file size using ``self.model_size`` inside LightningModule.
```
Comment (SkafteNicki) [inline docs/source/common/lightning_module.rst:1010]: what is the unit of this thing?
```suggestion
Get the model file size (in megabytes) using ``self.model_size`` inside LightningModule.
```

File: docs/source/common/lightning_module.rst [hunk 13]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
...
+
+        loss = ...
+        self.manual_backward(loss, opt)
+        opt.step()
+        opt.zero_grad()
+
+This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers
+AND you are an expert user. Most useful for research like reinforcement learning, sparse coding and GAN research.
+
+In the multi-optimizer case, ignore the optimizer_idx flag and use the optimizers directly
```
Comment (SkafteNicki) [inline docs/source/common/lightning_module.rst:957]:
```suggestion
In the multi-optimizer case, ignore the ``optimizer_idx`` argument and use the optimizers directly
```

**Review by SkafteNicki — APPROVED** [review]
LGTM

File: docs/source/common/lightning_module.rst (hunk 1 above)
Comment (Borda) [reply docs/source/common/lightning_module.rst:935]: @tchaton ^^

**Review by Borda — APPROVED** [review]

File: README.md (hunk 7 above)
Comment (carmocca) [reply README.md:323]: correct, the comment should be removed

File: README.md [hunk 14]
```diff
@@ -315,6 +315,10 @@ For complex/professional level work, you have optional full control of the train
 
 ```python
 class LitAutoEncoder(pl.LightningModule):
+    def __init__(self):
+        super().__init__()
+        self.automatic_optimization = False
+
```
Comment (carmocca) [inline README.md:321]: Can you use a `@property instead`? and lets put it below training_step

File: docs/source/common/lightning_module.rst [hunk 15]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
...
 
+automatic_optimization
+~~~~~~~~~~~~~~~~~~~~~~
+When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for your own
+optimizer behavior
+
+.. code-block:: python
+
+    def __init__(self):
+        self.automatic_optimization = False
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:943]: Same here

File: docs/source/common/lightning_module.rst [hunk 16]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
 
 --------------
 
+automatic_optimization
+~~~~~~~~~~~~~~~~~~~~~~
+When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for your own
+optimizer behavior
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:938]:
```suggestion
When set to ``False``, Lightning does not automate the optimization process. This means you are responsible for handling your optimizers. However, we do take care of precision and any accelerators used.
```

File: docs/source/common/lightning_module.rst [hunk 17]
```diff
@@ -932,9 +932,88 @@ True if using TPUs
...
+        # access your optimizers with use_pl_optimizer=False. Default is True
+        opt = self.optimizers(use_pl_optimizer=True)
+
+        loss = ...
+        self.manual_backward(loss, opt)
+        opt.step()
+        opt.zero_grad()
+
+This is not recommended when using a single optimizer, instead it's recommended when using 2+ optimizers
+AND you are an expert user. Most useful for research like reinforcement learning, sparse coding and GAN research.
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:955]:
```suggestion
This is recommended only if using 2+ optimizers AND if you know how to perform the optimization procedure properly. Note that automatic optimization can still be used with multiple optimizers by relying on the ``optimizer_idx`` parameter. Manual optimization is most useful for research topics like reinforcement learning, sparse coding, and GAN research.
```

File: docs/source/common/lightning_module.rst (hunk 2 above)
Comment (carmocca) [reply docs/source/common/lightning_module.rst:976]: We use zero_grad after step throught Lightning, however, recently we have noticed that GANs can fail to learn if it is not done before backward. see: #5545

File: README.md (hunk 14 above)
Comment (rohitgr7) [reply README.md:321]: it's already a property in LightningModule.

Comment (carmocca) [reply README.md:321]: Ah it's a getter/setter in the LightningModule... I thought it was just a simple getter like the one I wrote.

Not sure what's the advantage of the setter since it's a simple flag

**Comment by carmocca** [comment]
Please, do not merge until all comments are resolved :)

File: README.md (hunk 14 above)
Comment (awaelchli) [reply README.md:321]: I see two advantages:
1) can attach docs with a docstring
2) can implement design by contract if needed, and responsibility of updating the state lies in the hands of the LightningModule not the external modifier.

2) granted, is not so relevant here in this trivial setter, but I still see value

File: README.md (hunk 7 above)
Comment (rohitgr7) [reply README.md:323]: done.

File: docs/source/common/lightning_module.rst (hunk 2 above)
Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:973]: yep makes sense.. since there might be some left over gradients in the generator here from `disc_loss.backward()`

**Review by SeanNaren — APPROVED** [review]

File: docs/source/common/lightning_module.rst [hunk 18]
```diff
@@ -841,7 +841,7 @@ The current step (does not reset each epoch)
 
 hparams
 ~~~~~~~
-After calling `save_hyperparameters` anything passed to init() is available via hparams.
+After calling ``save_hyperparameters`` anything passed to ``__init__()`` is available via the ``hparams`` attribute.
```
Comment (kaushikb11) [inline docs/source/common/lightning_module.rst:844]: It's not anything passed, right? @rohitgr7 

```suggestion
The arguments saved by calling ``save_hyperparameters`` passed through ``__init__()`` could be accessed by the ``hparams`` attribute.
```

Comment (rohitgr7) [reply docs/source/common/lightning_module.rst:844]: :+1: ... can you update this in your PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/6056??

Comment (kaushikb11) [reply docs/source/common/lightning_module.rst:844]: @Borda Updated it here https://github.com/PyTorchLightning/pytorch-lightning/pull/6065 :)
